### PR TITLE
Use *-sys crates where possible

### LINKED ISSUE #514: Please support using existing Rust library crates for zstd, zlib, etc
The rocksdb crate only seems to support vendoring zstd/zlib/etc, which means if a Rust crate depends on rocksdb and also pulls in zstd/zlib/etc, it'll end up with symbol conflicts between the rocksdb vendored zstd/zlib/etc and the other version they include.

Please consider using crates like `zstd-sys` and `libz-sys` rather than vendoring. Those crates do support statically vendoring the library, and also support not doing so, giving the user of rocksdb control over how they want to use zstd/zlib/etc.

### LINKED ISSUE #535: linking error: multiple definition of `ZSTD_xxx in release mode
I  am getting linking error in release mode for multiple version of zstd library.  In addition to lobrocksdb_sys, async_compression also uses zstd.  It works fine in debug mode.

If I pass `RUSTFLAGS="-Clink-arg=-Wl,--allow-multiple-definition"`, it works fine but that's workaround.

```
 rustc --version
rustc 1.55.0-nightly (7c3872e6b 2021-06-24)
```
```
 cc --version
cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
```

Here is my Cargo.lock file entries for rocksdb and zstd.

```
[[package]]
name = "rocksdb"
version = "0.15.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "23d83c02c429044d58474eaf5ae31e062d0de894e21125b47437ec0edc1397e6"
dependencies = [
 "libc",
 "librocksdb-sys",
]

[[package]]
name = "librocksdb-sys"
version = "6.17.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5da125e1c0f22c7cae785982115523a0738728498547f415c9054cb17c7e89f9"
dependencies = [
 "bindgen",
 "cc",
 "glob",
 "libc",
]

[[package]]
name = "async-compression"
version = "0.3.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5443ccbb270374a2b1055fc72da40e1f237809cd6bb0e97e66d264cd138473a6"
dependencies = [
 "bytes 0.5.6",
 "futures-core",
 "memchr 2.4.0",
 "pin-project-lite",
 "tokio",
 "zstd",
 "zstd-safe",
]

[[package]]
name = "zstd"
version = "0.7.0+zstd.1.4.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9428752481d8372e15b1bf779ea518a179ad6c771cca2d2c60e4fbff3cc2cd52"
dependencies = [
 "zstd-safe",
]
[[package]]
name = "zstd-safe"
version = "3.1.0+zstd.1.4.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5aa1926623ad7fe406e090555387daf73db555b948134b4d73eac5eb08fb666d"
dependencies = [
 "libc",
 "zstd-sys",
]
[[package]]
name = "zstd-sys"
version = "1.5.0+zstd.1.4.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4e6c094340240369025fc6b731b054ee2a834328fa584310ac96aa4baebdc465"
dependencies = [
 "cc",
 "libc",
]n
```
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "/tmp/xxx-ws-async/target/release/deps/xxxx_db-3521ae0ffe672ee6.xxxx_db.7wvb29c8-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/tmp/xxx-ws-async/target/release/deps" "-L" "/tmp/kdb-ws-async/target/release/build/zstd-sys-9aa6a79a74378f28/out" "-L" "/tmp/kdb-ws-async/target/release/build/librocksdb-sys-3ad2f1387ab4c043/out" "-L" "/tmp/kdb-ws-async/target/release/build/librocksdb-sys-3ad2f1387ab4c043/out" "-L" "/tmp/kdb-ws-async/target/release/build/librocksdb-sys-3ad2f1387ab4c043/out" "-L" "/tmp/kdb-ws-async/target/release/build/librocksdb-sys-3ad2f1387ab4c043/out" "-L" "/tmp/kdb-ws-async/target/release/build/librocksdb-sys-3ad2f1387ab4c043/out" "-L" "/tmp/kdb-ws-async/target/release/build/librocksdb-sys-3ad2f1387ab4c043/out" "-L" "/usr/lib" "-L" "/tmp/kdb-ws-async/target/release/build/sys-info-c195f27c66ea515b/out" "-L" "/tmp/kdb-ws-async/target/release/build/libmimalloc-sys-7bec45b781c0a3bd/out" "-L" "/root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/tmp/rustce7rXpn/libsys_info-b5c54aab1007cd9e.rlib" "/tmp/rustce7rXpn/libzstd_sys-4eb8990bed6a18be.rlib" "/tmp/rustce7rXpn/liblibrocksdb_sys-6d6326e3337d8085.rlib" "/tmp/rustce7rXpn/liblibmimalloc_sys-59db1a092107eaf4.rlib" "-Wl,--start-group" "-Wl,--end-group" "/root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-d231748acf961207.rlib" "-Wl,-Bdynamic" "-lstdc++" "-lpython3.6m" "-lz" "-lssl" "-lcrypto" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/tmp/kdb-ws-async/target/release/deps/xxxx_db-3521ae0ffe672ee6" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs"
  = note: /tmp/rustce7rXpn/liblibrocksdb_sys-6d6326e3337d8085.rlib(zstd_common.o): In function `ZSTD_versionNumber':
          zstd_common.c:(.text.ZSTD_versionNumber+0x0): multiple definition of `ZSTD_versionNumber'
          /tmp/rustce7rXpn/libzstd_sys-4eb8990bed6a18be.rlib(zstd_common.o):zstd_common.c:(.text.ZSTD_versionNumber+0x0): first defined here
          /tmp/rustce7rXpn/liblibrocksdb_sys-6d6326e3337d8085.rlib(zstd_common.o): In function `ZSTD_versionString':
          zstd_common.c:(.text.ZSTD_versionString+0x0): multiple definition of `ZSTD_versionString'
          /tmp/rustce7rXpn/libzstd_sys-4eb8990bed6a18be.rlib(zstd_common.o):zstd_common.c:(.text.ZSTD_versionString+0x0): first defined here
          /tmp/rustce7rXpn/liblibrocksdb_sys-6d6326e3337d8085.rlib(zstd_common.o): In function `ZSTD_isError':
          zstd_common.c:(.text.ZSTD_isError+0x0): multiple definition of `ZSTD_isError'
          /tmp/rustce7rXpn/libzstd_sys-4eb8990bed6a18be.rlib(zstd_common.o):zstd_common.c:(.text.ZSTD_isError+0x0): first defined here
          /tmp/rustce7rXpn/liblibrocksdb_sys-6d6326e3337d8085.rlib(zstd_common.o): In function `ZSTD_getErrorName':
          zstd_common.c:(.text.ZSTD_getErrorName+0x0): multiple definition of `ZSTD_getErrorName'
          /tmp/rustce7rXpn/libzstd_sys-4eb8990bed6a18be.rlib(zstd_common.o):zstd_common.c:(.text.ZSTD_getErrorName+0x0): first defined here
          /tmp/rustce7rXpn/liblibrocksdb_sys-6d6326e3337d8085.rlib(zstd_common.o): In function `ZSTD_getErrorCode':
          zstd_common.c:(.text.ZSTD_getErrorCode+0x0): multiple definition of `ZSTD_getErrorCode'
          /tmp/rustce7rXpn/libzstd_sys-4eb8990bed6a18be.rlib(zstd_common.o):zstd_common.c:(.text.ZSTD_getErrorCode+0x0): first defined here
          /tmp/rustce7rXpn/liblibrocksdb_sys-6d6326e3337d8085.rlib(zstd_common.o): In function `ZSTD_getErrorString':
          zstd_common.c:(.text.ZSTD_getErrorString+0x0): multiple definition of `ZSTD_getErrorString'
          /tmp/rustce7rXpn/libzstd_sys-4eb8990bed6a18be.rlib(zstd_common.o):zstd_common.c:(.text.ZSTD_getErrorString+0x0): first defined here
          collect2: error: ld returned 1 exit status
error: aborting due to previous error
error: could not compile `xxx-db`
To learn more, run the command again with --verbose.
```

### PR DESCRIPTION
**Pull request by Dr-Emann**
Depend on `bzip2-sys`, `libz-sys`, and `zstd-sys`, rather than bundling them ourselves. This should reduce the burden of keeping those libraries up to date: someone else has already figured out how to do the build system, etc. Additionally, this avoids issues with multiple versions of the same functions when `rocksdb-sys` links against say `zstd`, and another crate pulls in `zstd-sys`, which also links against zstd.

This does not use `snappy-sys` or `lz4-sys` because neither currently exposes the include directory (via "cargo:include=..", which we can use as "DEP_XXX_INCLUDE") to dependencies when building a vendored version of the library, which we need.

Additionally, add tests for creating a DB with each compression type (checking that each compression type works/doesn't work as expected based on the features. Also put the zstd specific test behind a cfg, so tests pass with `--no-default-features`

Fixes #514 
Fixes #535
Fixes Nemo157/async-compression#127

### CODE REVIEW

**Comment by Dr-Emann** [comment]
Is there anything I can do to try to move this along?

File: librocksdb-sys/src/lib.rs [hunk 1]
```diff
@@ -17,6 +17,14 @@
 #![allow(non_snake_case)]
 #![allow(non_upper_case_globals)]
 
+// Ensure the libraries are linked in, despite it not being used directly
```
Comment (stanislav-tkach) [inline librocksdb-sys/src/lib.rs:20]: I'm not sure it works this way, but it looks good otherwise.

**Review by stanislav-tkach — APPROVED** [review]

File: librocksdb-sys/src/lib.rs (hunk 1 above)
Comment (Dr-Emann) [reply librocksdb-sys/src/lib.rs:20]: I'm not sure if it is or not, I think this is what made me think it was necessary: https://kornel.ski/rust-sys-crate#linking.

**Comment by Logarithmus** [comment]
@Dr-Emann please, resolve the conflicts
@aleksuss gentle ping

**Review by aleksuss — APPROVED** [review]
LGTM.

File: librocksdb-sys/src/lib.rs (hunk 1 above)
Comment (Dr-Emann) [reply librocksdb-sys/src/lib.rs:20]: Commenting one out does cause linking to fail with unknown symbols, so I think it's needed

**Comment by Dr-Emann** [comment]
I resolved the conflicts

**Comment by nemosupremo** [comment]
If I'm reading this right, then this means it's still not possible to have a project which depends on both `rust-rocksdb` and `lz4-sys`?

**Pull request merged**